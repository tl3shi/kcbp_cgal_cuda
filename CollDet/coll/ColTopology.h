
//***************************************************************************
//                              ColTopology
//***************************************************************************
//  Copyright (C):
//***************************************************************************
//CVSId: "@(#)$Id: ColTopology.h,v 1.5 2004/03/09 13:39:49 ehlgen Exp $"
//***************************************************************************


#ifndef ColTopology_H
#define ColTopology_H
#if defined(__sgi) || defined(_WIN32)
#pragma once
#endif

//---------------------------------------------------------------------------
//  Includes
//---------------------------------------------------------------------------

#include <vector>

#include <ColDopTree.h>					// just needed for constant NumOri
#include <ColUtils.h>					// just needed for constant NearZero
#include <col_import_export.h>


namespace col {

//---------------------------------------------------------------------------
//  Forward References
//---------------------------------------------------------------------------

//---------------------------------------------------------------------------
//   Constants
//---------------------------------------------------------------------------

//***************************************************************************
//  TopoFace
//***************************************************************************

struct COL_EXPORTIMPORT TopoFace
{
	///  A face is just an array of indices into some vertex array
	std::vector<unsigned int> v;

	TopoFace( const unsigned int vertex_indices[], unsigned int size );
	TopoFace( const std::vector< unsigned int > &v );
	TopoFace( const TopoFace &source );
	TopoFace( void );
	void operator = ( const TopoFace &source );
	void set( const unsigned int vertex_indices[], unsigned int size );

	void print( void ) const;

	unsigned int & operator [] ( int index );
	unsigned int operator [] ( int index ) const;
	unsigned int size( void ) const;
	void resize( unsigned int newsize );

};



//***************************************************************************
//  Topology
//***************************************************************************

class COL_EXPORTIMPORT Topology
{

public:

//---------------------------------------------------------------------------
//  Public Types
//---------------------------------------------------------------------------

	typedef std::vector<unsigned int>		FaceNeighbors;
	typedef FaceNeighbors::const_iterator	FaceNeighborIterator;
	typedef std::vector<unsigned int>		VertexNeighbors;
	typedef VertexNeighbors::const_iterator	VertexNeighborIterator;

//---------------------------------------------------------------------------
//  Public Instance methods
//---------------------------------------------------------------------------

	Topology( );
	explicit Topology( const Topology & source );
	void operator = ( const Topology & source );

	Topology( const std::vector<TopoFace> &face );
	Topology( const unsigned int face_a[][Dop::NumOri],
			  unsigned int nfaces,
			  unsigned int face_nv[] );
	Topology( const osg::GeometryPtr geom,
			  bool unify = true, float tolerance = NearZero );
	void operator = ( const std::vector<TopoFace> &face );
	void createFromGeom( const osg::GeometryPtr geom,
						 bool unify = false, float tolerance = NearZero );
	void create( const unsigned int face_a[][Dop::NumOri],
				 unsigned int nfaces, unsigned int face_nv[] );

	unsigned int v_neighbors	( unsigned int v_index ) const;
	unsigned int v_degree		( unsigned int v_index ) const;
	unsigned int v2f_size		( unsigned int v_index ) const;
	unsigned int f_size			( unsigned int f_index_in ) const;
    const unsigned int *f_index ( unsigned int f_index_in ) const;

	const std::vector<TopoFace>	&getFaceVector( void ) const;

	Topology::VertexNeighborIterator
		vertexNeighborBegin( unsigned int v_index ) const;
	Topology::VertexNeighborIterator
		vertexNeighborEnd( unsigned int v_index ) const;
	Topology::FaceNeighborIterator
		faceNeighborBegin( unsigned int f_index ) const;
	Topology::FaceNeighborIterator
		faceNeighborEnd( unsigned int f_index ) const;

    void setVertexEqualityEpsilon(OSG::Real32 epsilon);
    void setVertexEqualityTest(bool status);

	void print( void ) const;

	virtual ~Topology() throw();

//---------------------------------------------------------------------------
//  Public Class methods
//---------------------------------------------------------------------------

//---------------------------------------------------------------------------
//  Instance variables
//---------------------------------------------------------------------------

protected:

	/** vertex to face incidence relation; see createNeighbors();
	 *  v2f[i][j] is an index of a face that contains vertex i .
	 */
	std::vector< std::vector<unsigned int> >	m_v2f;

	/** vertex to vertex adjacency relation; see createNeighbors();
	 *  v2v[i][j] is an index of a vertex that is adjacent to i .
	 */
	std::vector<VertexNeighbors>				m_v2v;

	/** vertex to vertex adjacency relation; see createNeighbors();
	 *  f2f[i][j] is an index of a face that is adjacent to i .
	 */
	std::vector<FaceNeighbors>					m_f2f;

    /// face to vertex incidence relation (the array of faces)
    std::vector<TopoFace>                       m_f2v;

	/** Equity relation of vertex indices.
	 *
	 *  vEequality[i] is a vertex index that should be treated equal to 
	 *  vertex i .
	 *  If this vector is empty, then vertices have not been unified.
	 */
    std::vector<unsigned int>					m_vEquivClass;

//---------------------------------------------------------------------------
//  Private Instance methods
//---------------------------------------------------------------------------

protected:

	void createRelations( void );

};



//***************************************************************************
//  VertexIterator
//***************************************************************************

class COL_EXPORTIMPORT VertexIterator
{
public:

	VertexIterator( const Topology &topo, unsigned int v_index );
	VertexIterator( const VertexIterator &source );
	void operator = ( const VertexIterator &source );

	void operator ++ ( void );
	unsigned int operator * ( ) const;
	bool operator == ( const VertexIterator &other ) const;
	bool operator != ( const VertexIterator &other ) const;

protected:

	/// well ...
	const Topology * m_topo;

	/// the iterator will walk raound this vertex
	unsigned int m_v_index;

	/// the current neighbor
	unsigned int m_neighbor;

};


} // namespace col

#endif /* ColTopology_H */

