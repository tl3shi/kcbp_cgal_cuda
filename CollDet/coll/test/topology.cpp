
/*****************************************************************************\
 *                              topology
\*****************************************************************************/

/*! @file 
 *
 *  @brief
 *    Testprogram for class Topology.
 *
 *  @author Gabriel Zachmann
 *  
 *  @todo
 *    topology.out neu erzeugen, wenn OSG's Kugel in Ordnung ist, d.h.,
 *    dass Punkte nicht mehrfach vorkommen.
 */


//---------------------------------------------------------------------------
//  Includes
//---------------------------------------------------------------------------


#include <stdlib.h>
#include <stdio.h>

#include <ColTopology.h>
#include <ColUtils.h>

#include <OpenSG/OSGVector.h>
#include <OpenSG/OSGSimpleGeometry.h>


//**************************************************************************
// Utility functions
//**************************************************************************

enum GeometryTypeE { OBJ_PLANE, OBJ_SPHERE, OBJ_TORUS, OBJ_BOX };

osg::GeometryPtr createGeom( GeometryTypeE type, unsigned int complexity )
{
	osg::NodePtr node;
	osg::GeometryPtr geom;

    switch ( type )
    {
        case OBJ_PLANE:
            node = osg::makePlane( 1, 1, complexity, complexity );
            break;

        case OBJ_SPHERE:
            complexity = static_cast<int>( logf( complexity*complexity ) );
            node = osg::makeSphere( complexity, 0.5 );
            break;

        case OBJ_TORUS:
            node = osg::makeTorus( 0.35, 0.5, complexity+1, complexity+1 );
            break;

        case OBJ_BOX:
            node = col::makeCube( 0.5, GL_QUADS );
            break;

        default:
            fputs("createGeom: BUG: type is out of range!\n",stderr);
            exit(-1);
    }

    if ( node == osg::NullFC )
    {
        fprintf(stderr,"topology: createGeom failed!\n");
        exit(-1);                   // makes no sense to continue
    }

    geom  = osg::GeometryPtr::dcast( node->getCore() );

    if ( geom == osg::NullFC )
    {
        fprintf(stderr,"createGeom: dcast(GeometryPtr) returned NULL!\n");
        exit(-1);
    }

	return geom;
}


void printPoints( osg::GeometryPtr geom )
{
	const osg::MFPnt3f *points = col::getPoints( geom );
	for ( unsigned int i = 0; i < points->size(); i ++ )
		col::printPnt( (*points)[i] );
}




//**************************************************************************
// main
//**************************************************************************


int main( void )
{
	osg::osgInit( 0, NULL );

	osg::Pnt3f pnt[8] = { Pnt3f(-1,-1, 1),
						  Pnt3f( 1,-1, 1),
						  Pnt3f( 1, 1, 1),
						  Pnt3f(-1, 1, 1),
						  Pnt3f(-1,-1,-1),
						  Pnt3f( 1,-1,-1),
						  Pnt3f( 1, 1,-1),
						  Pnt3f(-1, 1,-1)
						};
    unsigned int face[6][col::Dop::NumOri] = { {0,1,2,3},
										  {1,5,6,2},
										  {2,6,7,3},
										  {0,3,7,4},
										  {0,4,5,1},
										  {4,7,6,5}
										};
    unsigned int face_nv[6] = { 4, 4, 4, 4, 4, 4 };

	col::Topology topo1( face, 6, face_nv );

	puts("\nCube from array:\n");
	puts("points:");
	for ( int i = 0; i < 8; i ++ )
		col::printPnt( pnt[i] );
	puts("topology:");
	topo1.print();

	osg::GeometryPtr geom;

	geom = createGeom( OBJ_PLANE, 3 );
	col::Topology topo2( geom );

	puts("\nPlane:\n");
	puts("points:");
	printPoints( geom );
	puts("topology:");
	topo2.print();

	geom = createGeom( OBJ_SPHERE, 1 );
	col::Topology topo3( geom );

	puts("\nSphere:\n");
	puts("points:");
	printPoints( geom );
	puts("topology:");
	topo3.print();

	geom = createGeom( OBJ_TORUS, 3 );
	col::Topology topo4( geom, true, 1E-4 );

	puts("\nTorus with vertex unification:\n");
	puts("points:");
	printPoints( geom );
	puts("topology:");
	topo4.print();

	geom = createGeom( OBJ_TORUS, 3 );
	topo4.createFromGeom( geom, false, 0 );

	puts("\nTorus without vertex unification:\n");
	puts("points:");
	printPoints( geom );
	puts("topology:");
	topo4.print();

	geom = createGeom( OBJ_BOX, 0 );
	col::Topology topo5( geom );		// does not do unification by default
	puts("\nBox:\n");
	puts("points:");
	printPoints( geom );
	puts("topology:");
	topo5.print();

	return 0;
}


char cvsid[] =
"@(#)$Id: topology.cpp,v 1.6 2004/03/09 13:41:36 ehlgen Exp $";



/**  Zum Abpasten fuer eine Funktion, Methode, oder Makro-"Funktion".
 *
 * @param param1	Comment for param1
 *
 * @return
 *   Text for return value.
 *
 * Detaillierte Beschreibung ...
 *
 * @throw Exception
 *   Beschreibung ..
 *
 * @warning
 *   Dinge, die der Aufrufer unbedingt beachten muss...
 *
 * @pre
 *   Annahmnen, die die Funktion macht...
 *
 * @sideeffects
 *   Nebenwirkungen, globale Variablen, die veraendert werden, ..
 *
 * @todo
 *   Was noch getan werden muss
 *
 * @bug
 *   Bekannte Bugs dieser Funktion
 *
 * @see
 *   ...
 *
 * @implementation
 *   Implementierungsdetails, TODOs, ...
 *
 **/




