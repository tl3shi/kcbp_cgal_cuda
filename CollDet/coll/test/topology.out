
Cube from array:

points:
-1.000000 -1.000000  1.000000
 1.000000 -1.000000  1.000000
 1.000000  1.000000  1.000000
-1.000000  1.000000  1.000000
-1.000000 -1.000000 -1.000000
 1.000000 -1.000000 -1.000000
 1.000000  1.000000 -1.000000
-1.000000  1.000000 -1.000000
topology:
     face to vertex:
  0:   0   1   2   3 
  1:   1   5   6   2 
  2:   2   6   7   3 
  3:   0   3   7   4 
  4:   0   4   5   1 
  5:   4   7   6   5 
     vertex to face:
  0:   0   3   4 
  1:   0   1   4 
  2:   0   1   2 
  3:   0   2   3 
  4:   3   4   5 
  5:   1   4   5 
  6:   1   2   5 
  7:   2   3   5 
     vertex to vertex:
  0:   3   1   4 
  1:   0   2   5 
  2:   1   3   6 
  3:   0   2   7 
  4:   0   7   5 
  5:   1   6   4 
  6:   5   2   7 
  7:   6   3   4 
     face to face:
  0:   3   4   1   2 
  1:   0   4   5   2 
  2:   0   1   5   3 
  3:   4   0   2   5 
  4:   0   3   5   1 
  5:   4   3   2   1 

Plane:

points:
-0.500000 -0.500000  0.000000
-0.166667 -0.500000  0.000000
 0.166667 -0.500000  0.000000
 0.500000 -0.500000  0.000000
-0.500000 -0.166667  0.000000
-0.166667 -0.166667  0.000000
 0.166667 -0.166667  0.000000
 0.500000 -0.166667  0.000000
-0.500000  0.166667  0.000000
-0.166667  0.166667  0.000000
 0.166667  0.166667  0.000000
 0.500000  0.166667  0.000000
-0.500000  0.500000  0.000000
-0.166667  0.500000  0.000000
 0.166667  0.500000  0.000000
 0.500000  0.500000  0.000000
topology:
     face to vertex:
  0:   4   0   5 
  1:   5   0   1 
  2:   5   1   6 
  3:   6   1   2 
  4:   6   2   7 
  5:   7   2   3 
  6:   8   4   9 
  7:   9   4   5 
  8:   9   5  10 
  9:  10   5   6 
 10:  10   6  11 
 11:  11   6   7 
 12:  12   8  13 
 13:  13   8   9 
 14:  13   9  14 
 15:  14   9  10 
 16:  14  10  15 
 17:  15  10  11 
     vertex to face:
  0:   0   1 
  1:   1   2   3 
  2:   3   4   5 
  3:   5 
  4:   0   6   7 
  5:   0   1   2   7   8   9 
  6:   2   3   4   9  10  11 
  7:   4   5  11 
  8:   6  12  13 
  9:   6   7   8  13  14  15 
 10:   8   9  10  15  16  17 
 11:  10  11  17 
 12:  12 
 13:  12  13  14 
 14:  14  15  16 
 15:  16  17 
     vertex to vertex:
  0:   4   5   1 
  1:   5   0   6   2 
  2:   6   1   7   3 
  3:   7   2 
  4:   5   0   8   9 
  5:   4   0   1   6   9  10 
  6:   5   1   2   7  10  11 
  7:   6   2   3  11 
  8:   9   4  12  13 
  9:   8   4   5  10  13  14 
 10:   9   5   6  11  14  15 
 11:  10   6   7  15 
 12:  13   8 
 13:  12   8   9  14 
 14:  13   9  10  15 
 15:  14  10  11 
     face to face:
  0:   7   1 
  1:   2   0 
  2:   9   1   3 
  3:   4   2 
  4:  11   3   5 
  5:   4 
  6:  13   7 
  7:   8   6   0 
  8:  15   7   9 
  9:  10   8   2 
 10:  17   9  11 
 11:  10   4 
 12:  13 
 13:  14  12   6 
 14:  13  15 
 15:  16  14   8 
 16:  15  17 
 17:  16  10 

Sphere:

points:
 0.500000  0.000000 -0.000000
 0.223607  0.000000 -0.447214
-0.223607  0.000000  0.447214
-0.500000  0.000000  0.000000
 0.223607  0.425325 -0.138197
-0.223607  0.425325  0.138197
 0.223607 -0.425325 -0.138197
-0.223607 -0.425325  0.138197
-0.223607  0.262866 -0.361803
 0.223607  0.262866  0.361803
-0.223607 -0.262866 -0.361803
 0.223607 -0.262866  0.361803
topology:
     face to vertex:
  0:   1   4   0 
  1:   4   9   0 
  2:   4   5   9 
  3:   8   5   4 
  4:   1   8   4 
  5:   1  10   8 
  6:  10   3   8 
  7:   8   3   5 
  8:   3   2   5 
  9:   3   7   2 
 10:   3  10   7 
 11:  10   6   7 
 12:   6  11   7 
 13:   6   0  11 
 14:   6   1   0 
 15:  10   1   6 
 16:  11   0   9 
 17:   2  11   9 
 18:   5   2   9 
 19:  11   2   7 
     vertex to face:
  0:   0   1  13  14  16 
  1:   0   4   5  14  15 
  2:   8   9  17  18  19 
  3:   6   7   8   9  10 
  4:   0   1   2   3   4 
  5:   2   3   7   8  18 
  6:  11  12  13  14  15 
  7:   9  10  11  12  19 
  8:   3   4   5   6   7 
  9:   1   2  16  17  18 
 10:   5   6  10  11  15 
 11:  12  13  16  17  19 
     vertex to vertex:
  0:   1   4   9   6  11 
  1:   0   4   8  10   6 
  2:   3   5   7   9  11 
  3:  10   8   5   2   7 
  4:   1   0   9   5   8 
  5:   4   9   8   3   2 
  6:  10   7  11   0   1 
  7:   3   2  10   6  11 
  8:   4   5   1  10   3 
  9:   4   0   5  11   2 
 10:   1   8   3   7   6 
 11:   6   7   0   9   2 
     face to face:
  0:  14   4   1 
  1:   0   2  16 
  2:   1   3  18 
  3:   4   7   2 
  4:   0   5   3 
  5:   4  15   6 
  6:   5  10   7 
  7:   3   6   8 
  8:   7   9  18 
  9:   8  10  19 
 10:   9   6  11 
 11:  10  15  12 
 12:  11  13  19 
 13:  12  14  16 
 14:  13  15   0 
 15:  11   5  14 
 16:  17  13   1 
 17:  18  19  16 
 18:   2   8  17 
 19:  12  17   9 

Torus with vertex unification:

points:
 0.850000 -0.000000  0.000000
 0.500000 -0.000000  0.350000
 0.150000 -0.000000 -0.000000
 0.500000 -0.000000 -0.350000
 0.850000 -0.000000  0.000000
-0.000000 -0.850000  0.000000
-0.000000 -0.500000  0.350000
-0.000000 -0.150000 -0.000000
-0.000000 -0.500000 -0.350000
-0.000000 -0.850000  0.000000
-0.850000  0.000000  0.000000
-0.500000  0.000000  0.350000
-0.150000  0.000000 -0.000000
-0.500000  0.000000 -0.350000
-0.850000  0.000000  0.000000
 0.000000  0.850000  0.000000
 0.000000  0.500000  0.350000
 0.000000  0.150000 -0.000000
 0.000000  0.500000 -0.350000
 0.000000  0.850000  0.000000
 0.850000 -0.000000  0.000000
 0.500000 -0.000000  0.350000
 0.150000 -0.000000 -0.000000
 0.500000 -0.000000 -0.350000
 0.850000 -0.000000  0.000000
topology:
     face to vertex:
  0:   0   1   5 
  1:   5   1   6 
  2:   5   6  10 
  3:  10   6  11 
  4:  10  11  15 
  5:  15  11  16 
  6:  15  16   0 
  7:   0  16   1 
  8:   1   2   6 
  9:   6   2   7 
 10:   6   7  11 
 11:  11   7  12 
 12:  11  12  16 
 13:  16  12  17 
 14:  16  17   1 
 15:   1  17   2 
 16:   2   3   7 
 17:   7   3   8 
 18:   7   8  12 
 19:  12   8  13 
 20:  12  13  17 
 21:  17  13  18 
 22:  17  18   2 
 23:   2  18   3 
 24:   3   0   8 
 25:   8   0   5 
 26:   8   5  13 
 27:  13   5  10 
 28:  13  10  18 
 29:  18  10  15 
 30:  18  15   3 
 31:   3  15   0 
     vertex to face:
  0:   0   6   7  24  25  31 
  1:   0   1   7   8  14  15 
  2:   8   9  15  16  22  23 
  3:  16  17  23  24  30  31 
  4: 
  5:   0   1   2  25  26  27 
  6:   1   2   3   8   9  10 
  7:   9  10  11  16  17  18 
  8:  17  18  19  24  25  26 
  9: 
 10:   2   3   4  27  28  29 
 11:   3   4   5  10  11  12 
 12:  11  12  13  18  19  20 
 13:  19  20  21  26  27  28 
 14: 
 15:   4   5   6  29  30  31 
 16:   5   6   7  12  13  14 
 17:  13  14  15  20  21  22 
 18:  21  22  23  28  29  30 
     vertex to vertex:
  0:   5   1  15  16   3   8 
  1:   0   5   6  16   2  17 
  2:   1   6   7  17   3  18 
  3:   2   7   8  18   0  15 
  4: 
  5:   0   1   6  10   8  13 
  6:   5   1  10  11   2   7 
  7:   6   2  11  12   3   8 
  8:   7   3  12  13   0   5 
  9: 
 10:   5   6  11  15  13  18 
 11:  10   6  15  16   7  12 
 12:  11   7  16  17   8  13 
 13:  12   8  17  18   5  10 
 14: 
 15:  10  11  16   0  18   3 
 16:  15  11   0   1  12  17 
 17:  16  12   1   2  13  18 
 18:  17  13   2   3  10  15 
     face to face:
  0:  25   7   1 
  1:   2   0   8 
  2:  27   1   3 
  3:   4   2  10 
  4:  29   3   5 
  5:   6   4  12 
  6:  31   5   7 
  7:   0   6  14 
  8:   1  15   9 
  9:  10   8  16 
 10:   3   9  11 
 11:  12  10  18 
 12:   5  11  13 
 13:  14  12  20 
 14:   7  13  15 
 15:   8  14  22 
 16:   9  23  17 
 17:  18  16  24 
 18:  11  17  19 
 19:  20  18  26 
 20:  13  19  21 
 21:  22  20  28 
 22:  15  21  23 
 23:  16  22  30 
 24:  17  31  25 
 25:  26  24   0 
 26:  19  25  27 
 27:  28  26   2 
 28:  21  27  29 
 29:  30  28   4 
 30:  23  29  31 
 31:  24  30   6 

Torus without vertex unification:

points:
 0.850000 -0.000000  0.000000
 0.500000 -0.000000  0.350000
 0.150000 -0.000000 -0.000000
 0.500000 -0.000000 -0.350000
 0.850000 -0.000000  0.000000
-0.000000 -0.850000  0.000000
-0.000000 -0.500000  0.350000
-0.000000 -0.150000 -0.000000
-0.000000 -0.500000 -0.350000
-0.000000 -0.850000  0.000000
-0.850000  0.000000  0.000000
-0.500000  0.000000  0.350000
-0.150000  0.000000 -0.000000
-0.500000  0.000000 -0.350000
-0.850000  0.000000  0.000000
 0.000000  0.850000  0.000000
 0.000000  0.500000  0.350000
 0.000000  0.150000 -0.000000
 0.000000  0.500000 -0.350000
 0.000000  0.850000  0.000000
 0.850000 -0.000000  0.000000
 0.500000 -0.000000  0.350000
 0.150000 -0.000000 -0.000000
 0.500000 -0.000000 -0.350000
 0.850000 -0.000000  0.000000
topology:
     face to vertex:
  0:   0   1   5 
  1:   5   1   6 
  2:   5   6  10 
  3:  10   6  11 
  4:  10  11  15 
  5:  15  11  16 
  6:  15  16  20 
  7:  20  16  21 
  8:   1   2   6 
  9:   6   2   7 
 10:   6   7  11 
 11:  11   7  12 
 12:  11  12  16 
 13:  16  12  17 
 14:  16  17  21 
 15:  21  17  22 
 16:   2   3   7 
 17:   7   3   8 
 18:   7   8  12 
 19:  12   8  13 
 20:  12  13  17 
 21:  17  13  18 
 22:  17  18  22 
 23:  22  18  23 
 24:   3   4   8 
 25:   8   4   9 
 26:   8   9  13 
 27:  13   9  14 
 28:  13  14  18 
 29:  18  14  19 
 30:  18  19  23 
 31:  23  19  24 
     vertex to face:
  0:   0 
  1:   0   1   8 
  2:   8   9  16 
  3:  16  17  24 
  4:  24  25 
  5:   0   1   2 
  6:   1   2   3   8   9  10 
  7:   9  10  11  16  17  18 
  8:  17  18  19  24  25  26 
  9:  25  26  27 
 10:   2   3   4 
 11:   3   4   5  10  11  12 
 12:  11  12  13  18  19  20 
 13:  19  20  21  26  27  28 
 14:  27  28  29 
 15:   4   5   6 
 16:   5   6   7  12  13  14 
 17:  13  14  15  20  21  22 
 18:  21  22  23  28  29  30 
 19:  29  30  31 
 20:   6   7 
 21:   7  14  15 
 22:  15  22  23 
 23:  23  30  31 
 24:  31 
     vertex to vertex:
  0:   5   1 
  1:   0   5   6   2 
  2:   1   6   7   3 
  3:   2   7   8   4 
  4:   3   8   9 
  5:   0   1   6  10 
  6:   5   1  10  11   2   7 
  7:   6   2  11  12   3   8 
  8:   7   3  12  13   4   9 
  9:   8   4  13  14 
 10:   5   6  11  15 
 11:  10   6  15  16   7  12 
 12:  11   7  16  17   8  13 
 13:  12   8  17  18   9  14 
 14:  13   9  18  19 
 15:  10  11  16  20 
 16:  15  11  20  21  12  17 
 17:  16  12  21  22  13  18 
 18:  17  13  22  23  14  19 
 19:  18  14  23  24 
 20:  15  16  21 
 21:  20  16  17  22 
 22:  21  17  18  23 
 23:  22  18  19  24 
 24:  23  19 
     face to face:
  0:   1 
  1:   2   0   8 
  2:   1   3 
  3:   4   2  10 
  4:   3   5 
  5:   6   4  12 
  6:   5   7 
  7:   6  14 
  8:   1   9 
  9:  10   8  16 
 10:   3   9  11 
 11:  12  10  18 
 12:   5  11  13 
 13:  14  12  20 
 14:   7  13  15 
 15:  14  22 
 16:   9  17 
 17:  18  16  24 
 18:  11  17  19 
 19:  20  18  26 
 20:  13  19  21 
 21:  22  20  28 
 22:  15  21  23 
 23:  22  30 
 24:  17  25 
 25:  26  24 
 26:  19  25  27 
 27:  28  26 
 28:  21  27  29 
 29:  30  28 
 30:  23  29  31 
 31:  30 

Box:

points:
-0.500000 -0.500000  0.500000
 0.500000 -0.500000  0.500000
 0.500000  0.500000  0.500000
-0.500000  0.500000  0.500000
-0.500000 -0.500000 -0.500000
 0.500000 -0.500000 -0.500000
 0.500000  0.500000 -0.500000
-0.500000  0.500000 -0.500000
topology:
     face to vertex:
  0:   0   1   2   3 
  1:   1   5   6   2 
  2:   2   6   7   3 
  3:   0   3   7   4 
  4:   0   4   5   1 
  5:   4   7   6   5 
     vertex to face:
  0:   0   3   4 
  1:   0   1   4 
  2:   0   1   2 
  3:   0   2   3 
  4:   3   4   5 
  5:   1   4   5 
  6:   1   2   5 
  7:   2   3   5 
     vertex to vertex:
  0:   3   1   4 
  1:   0   2   5 
  2:   1   3   6 
  3:   0   2   7 
  4:   0   7   5 
  5:   1   6   4 
  6:   5   2   7 
  7:   6   3   4 
     face to face:
  0:   3   4   1   2 
  1:   0   4   5   2 
  2:   0   1   5   3 
  3:   4   0   2   5 
  4:   0   3   5   1 
  5:   4   3   2   1 
